File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.cpp
class Solution {
public:
    const int MOD = 1e9 + 7;
    
//     int solve(int ind, int state) {
//         if (ind == N) return state == 0;
//         if (ind > N) return 0;
//         int &ret = dp[ind][state];
//         if (ret != -1) return ret;
//         ret = 0;
//         if (state == 0) {
//             ret = solve(ind + 1, 0);
//             ret = (ret + (2*solve(ind + 1, 1)) % MOD)%MOD;
//         } else {
//             ret = (ret + solve(ind + 1, 1))%MOD;
//         }
//         ret = (ret + solve(ind + 2, 0)) % MOD;
//         return ret;
//     }
            
    int numTilings(int n) {
        dp.resize(3, vector<int>(2, 0));
        N = n;
        dp[N%3][0] = 1;
        for (int i = N - 1; i >= 0; i--) {
            dp[i%3][0] = ((dp[(i+1)%3][0] + dp[(i + 2)%3][0])%MOD + (2*dp[(i + 1)%3][1]) % MOD)%MOD;
            dp[i%3][1] = (dp[(i + 2)%3][0] + dp[(i + 1)%3][1]) % MOD;
        }
        return dp[0][0];
        // return solve(0, 0);
    }
private:
    int N;
    vector<vector<int>> dp;
};

File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.java
class Solution {
    int MOD = 1000000007;
    int solve(int ind, int msk, int[][] dp) {
        if (ind == -1) {
            if (msk == 0) return 1;
            else return 0;
        }
        int ret = dp[ind][msk];
        if (ret != -1) return ret;
        ret = 0;
        if (msk == 0) {
            ret = (ret + solve(ind - 1, 0, dp)) % MOD;
            ret = (ret + solve(ind - 1, 3, dp)) % MOD;
            ret = (ret + solve(ind - 1, 1, dp)) % MOD;
            ret = (ret + solve(ind - 1, 2, dp)) % MOD;
            
        } else if (msk == 1) {
            ret = (ret + solve(ind - 1, 2, dp)) % MOD;
            ret = (ret + solve(ind - 1, 3, dp)) % MOD;
        } else if (msk == 2) {
            ret = (ret + solve(ind - 1, 1, dp)) % MOD;
            ret = (ret + solve(ind - 1, 3, dp)) % MOD;
        } else {
            ret = solve(ind - 1, 0, dp) % MOD;
        }
        dp[ind][msk] = ret;
        return ret;
    }

    public int numTilings(int n) {
        int[][] dp = new int[n][4];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 4; j++) {
                dp[i][j] = -1;
            }
        }
        return solve(n-1, 0, dp);
    }
}